hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java

```diff
@@ -695,6 +695,7 @@ void sawNoSuchFamily() {
   private RegionCoprocessorHost coprocessorHost;
 
   private TableDescriptor htableDescriptor = null;
+  private ThreadPoolExecutor storeFileOpenAndCloseThreadPool;
```

Comment:
Why is this shared at the Region level. Shouldn't it be at the Store level (in HStore)?

What does sharing at the region level gain us? Are you attempting to evenly round-robin store open work over all opening stores in the region? Just sharing an executor at region level won't do this. If the underlying stores are skewed the order in which runnables are submitted to the executor will share the skew.

Reply:
If I understand the jira correctly, what you are trying to solve is below case.
One region with say 2 stores.  Store1 having much more files than other.  Say the config for the #threads in open pool is 10.  Now it will create 2 pools for each store with 5 threads each.  The Store2 will get finished soon.  But store1 will take much longer.    So if it was a shared pool of 10 threads the overall time for opening both stores would have been lesser.  my understanding correct?

Reply:
> If I understand the jira correctly, what you are trying to solve is below case.
> One region with say 2 stores. Store1 having much more files than other. Say the config for the #threads in open pool is 10. Now it will create 2 pools for each store with 5 threads each. The Store2 will get finished soon. But store1 will take much longer. So if it was a shared pool of 10 threads the overall time for opening both stores would have been lesser. my understanding correct?

Yes, just that case.

Reply:
> Why is this shared at the Region level. Shouldn't it be at the Store level (in HStore)?
> 
> What does sharing at the region level gain us? Are you attempting to evenly round-robin store open work over all opening stores in the region? Just sharing an executor at region level won't do this. If the underlying stores are skewed the order in which runnables are submitted to the executor will share the skew.

If only one store is too slow and that is the bottleneck of the whole open or close region process, it seems not easy to fix that by threadpool here(maybe hedged read?). As Anoops said above, the problem we want to solve here is some stores may have too many store files, while others have much fewer. It will be quicker ideally if  all of the store files just share the whole thread pool in one region.